File: LeetCode/Hao/120Triangle.java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        
        if (triangle.size() == 1) {
            return triangle.get(triangle.size() - 1).get(0);
        }
        
        int rowMax = triangle.size();
        int colMax = triangle.get(rowMax - 1).size();
        int[] d = new int[rowMax];
        
        
        ListIterator<List<Integer>> rowIterator = triangle.listIterator(rowMax);
        // initialize the dp 
        int i = 0;
        for (Integer data: rowIterator.previous()) {
            d[i] = data;
            i ++;
        }
        
        while (rowIterator.hasPrevious() == true) {
            List<Integer> rowList = rowIterator.previous();
            i = 0;
            
            for (Integer data:rowList) {
                d[i] = data + Math.min(d[i], d[i + 1]);
                i ++;
            }
        }
        
        return d[0];
    }
}

File: LeetCode/Lei/120Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int row = int(triangle.size());
        if(row==0)
          return 0;
        int sum[row][row];
        for(int i=0; i<row; i++)
          sum[row-1][i]=triangle[row-1][i];
        for(int i=row-2; i>=0; i--)
          for(int j=0; j<=i; j++)
            sum[i][j]=min(sum[i+1][j],sum[i+1][j+1])+triangle[i][j];
    return sum[0][0];
    }
};
